### DISCUSSION TITLE
torch_npu Skill 算子精度验证报告 - 环境兼容性问题发现

### DISCUSSION THREAD

**Discussion by zhangtao0408** [ascend-ai-coding/awesome-ascend-skills#35] [Skills-Verfication]
## 验证概述

本报告记录了通过多个并行 subagent 对 `torch_npu` skill 进行的算子精度验证过程和结果。

**验证服务器**: 175.99.1.3  
**验证日期**: 2025-03-02  
**分支**: feat/refactor-torch-npu-skill  
**最新提交**: 0e87a47

---

## 验证方法

启动了 **5 个并行 subagent** 分别验证不同类型的算子：

| Agent | 验证类型 | 算子列表 |
|-------|----------|----------|
| Agent 1 | 基础算子 | add, sub, mul, div |
| Agent 2 | 矩阵算子 | matmul, bmm, mm |
| Agent 3 | 激活函数 | relu, gelu, sigmoid, tanh |
| Agent 4 | 归一化算子 | layer_norm, softmax |
| Agent 5 | 注意力算子 | manual_attention |

---

### 服务器配置
```
NPU 设备: 8x Ascend 910B3
npu-smi 版本: 25.3.rc1
操作系统: Linux (aarch64)
```

### 软件版本
```
PyTorch: 2.8.0+cpu
torch_npu: 2.8.0.post1
CANN (NNAE): 8.2.RC1
Python: 3.11
```

---

### ⚠️ 发现环境兼容性问题

**所有算子测试均因环境配置问题失败**

#### 错误信息
```
error code: 561103
EZ9999: Parse dynamic kernel config fail.
Op MatMulV2 does not has any binary.
```

#### 详细结果

| 类别 | 算子 | 状态 | 错误类型 |
|------|------|------|----------|
| 基础算子 | torch.add | ❌ 失败 | aclnnAdd failed (561103) |
| 基础算子 | torch.sub | ❌ 失败 | aclnnSub failed (561103) |
| 基础算子 | torch.mul | ❌ 失败 | 精度超限 (abs=7.95e+00) |
| 基础算子 | torch.div | ❌ 失败 | 精度超限 (abs=5.43e+03) |
| 矩阵算子 | torch.matmul | ❌ 失败 | aclnnMatmul failed (561103) |
| 矩阵算子 | torch.bmm | ❌ 失败 | aclnnBatchMatMul failed (561103) |
| 矩阵算子 | torch.mm | ❌ 失败 | aclnnMm failed (561103) |
| 激活函数 | F.relu | ❌ 失败 | aclnnRelu failed (561103) |
| 激活函数 | F.gelu | ❌ 失败 | aclnnGelu failed (561103) |
| 激活函数 | F.sigmoid | ❌ 失败 | aclnnSigmoid failed (561103) |
| 激活函数 | F.tanh | ❌ 失败 | aclnnTanh failed (561103) |
| 归一化 | F.layer_norm | ❌ 失败 | aclnnInplaceOne failed (561103) |
| 归一化 | F.softmax | ❌ 失败 | 精度超限 (abs=3.43e+00) |
| 注意力 | manual_attention | ❌ 失败 | matmul 失败 |

**总计**: 0/14 通过

---

### 根本原因

**CANN 8.2.RC1 (NNAE) 与 torch_npu 2.8.0.post1 存在版本兼容性问题**

1. 错误码 `561103` 表示内核解析失败
2. NPU 硬件正常识别，但算子库缺失或不兼容
3. "Op does not has any binary" 表明缺少算子二进制文件

### 影响范围

- 所有 NPU 算子调用失败
- 部分算子回退到 CPU 导致精度差异巨大

---

## 对 skill 文档的建议改进

基于验证发现，建议在 `torch_npu` skill 中增强以下内容：

### 1. 版本配套检查

在 SKILL.md 中强调执行精度验证前必须确认版本配套：

```python
# 验证前检查版本配套
import torch
import torch_npu

print(f"PyTorch: {torch.__version__}")
print(f"torch_npu: {torch_npu.__version__}")

# 必须确认 CANN 版本与 torch_npu 版本配套
# 参见: references/installation/version_compatibility.md
```

### 2. 错误码参考

添加常见错误码说明：

| 错误码 | 含义 | 解决方案 |
|--------|------|----------|
| 561103 | 内核解析失败 | 检查 CANN/torch_npu 版本配套 |
| 561000 | 算子不存在 | 升级 CANN 或使用 op-plugin |

### 3. 环境配置步骤

在 `verification_guide.md` 中添加环境检查步骤：

```bash
# 1. 设置 CANN 环境
source /usr/local/Ascend/nnae/set_env.sh

# 2. 验证 NPU 可用
python3 -c \"import torch; import torch_npu; print(torch.npu.is_available())\"

# 3. 如返回 False 或报错，检查版本配套表
```

---

## 结论

1. **Skill 文档结构正确**：SKILL.md 和 references 文件结构完整，链接有效
2. **发现环境问题**：服务器 CANN/torch_npu 版本不配套导致算子无法执行
3. **建议改进**：在 skill 中增加更详细的版本检查和错误排查指导

---

## 相关链接

- PR: #33
- Issue: #32
- 前次验证报告: #34
